### ISSUE TITLE
Implement two qubit unitary to operations algorithm from https://arxiv.org/abs/quant-ph/0406176

### ISSUE THREAD

**Issue by NoureldinYosri** [quantumlib/Cirq#6777]
The paper https://arxiv.org/abs/quant-ph/0406176 introduces an algorithm for performing quantum shannon decomposition. We have this algorithm implemented in Cirq in https://github.com/quantumlib/Cirq/blob/351a08e52b7090cfab3a1ad07859ebf09d54052a/cirq-core/cirq/transformers/analytical_decompositions/quantum_shannon_decomposition.py#L44

However, one of the base cases of the algorithms is not implemented. Namely A.2 the base case for a two qubit unitary, instead cirq uses another decomposition https://github.com/quantumlib/Cirq/blob/351a08e52b7090cfab3a1ad07859ebf09d54052a/cirq-core/cirq/transformers/analytical_decompositions/quantum_shannon_decomposition.py#L101-L113

Which is not as efficient as the decompositin described in A.2. It would be nice to implement A.2 and uses it in QSD.

**What is the urgency from your perspective for this issue? Is it blocking important work?**

P2 - we should do it in the next couple of quarters

**Comment by RahilJain1366** [comment]
Hi @NoureldinYosri, can I take this up too since I am trying to find a solution for #6770?

**Comment by senecameeks** [comment]
Cirq Cynq: This will improve the performance by reducing the number of rotations and depth of the circuit.

**Comment by senecameeks** [comment]
@RahilJain1366, thanks for offering, let's discuss after #6770 is complete

**Comment by ldi18** [comment]
Hi, I’m also interested in working on this if @RahilJain1366 hasn’t taken it up yet @senecameeks .

**Comment by codrut3** [comment]
I implemented the A.2 case. However, when comparing unoptimized circuit depth for the current Cirq package with my change, I didn't observe any improvement:
```
qubits = [3, 4, 5, 6, 7, 8]
cirq_package = [45, 219, 963, 4035, 16515, 66819]
with_my_change = [47, 221, 965, 4037, 16517, 66821]
```
To obtain these numbers I generated a random unitary for the given number of qubits and then ran Shannon decomposition.

Note that these numbers are the same [as those here](https://github.com/quantumlib/Cirq/pull/6756#issuecomment-2404537243).

The reason for this is that the 3-qubit case already [uses the A.2 optimization](https://github.com/quantumlib/Cirq/blob/f3410e2174972a4ba05567732039f14811fcf5ca/cirq-core/cirq/transformers/analytical_decompositions/three_qubit_decomposition.py#L190). The crux of A.2 was already implemented by @balopat [in this method](https://github.com/quantumlib/Cirq/blob/f3410e2174972a4ba05567732039f14811fcf5ca/cirq-core/cirq/transformers/analytical_decompositions/two_qubit_to_cz.py#L89). Removing the 3-qubit base case and using A.2 for `n=2` doesn't help.

However, when running the optimizer:
`cirq.optimize_for_target_gateset(circuit, gateset=cirq.CZTargetGateset(preserve_moment_structure=False), max_num_passes=None)`
I observed a substantial improvement in the current package (without A.2 base case):
```
qubits = [3, 4, 5, 6, 7, 8]
cirq_package = [41, 203, 899, 3779, 15491, 62723]
```
which is better than [what was reported here](https://github.com/quantumlib/Cirq/pull/6756#issuecomment-2405821022).

Furthermore, I discovered that quantum Shannon decomposition is very slow. I traced this issue back to [`cleanup_operations`](https://github.com/quantumlib/Cirq/blame/f3410e2174972a4ba05567732039f14811fcf5ca/cirq-core/cirq/transformers/analytical_decompositions/two_qubit_to_cz.py#L185) in `two_qubit_to_cz.py`. `merge_single_qubit_gates_to_phased_x_and_z` is very slow, even for a small set of gates (size 7). I was able to optimize this substantially, reducing the time almost in half.

```
qubits = [3, 4, 5, 6, 7, 8]
time_sec_no_optimization = [0.05, 0.18, 0.76, 2.81, 11.22, 42.79]
time_sec_with_optimization = [0.04, 0.1, 0.39, 1.31, 5.43, 21.8]
```

You can see [my code here](https://github.com/codrut3/Cirq/tree/issue-6777).

To summarize:

* there is no need for the A.2 case as it's already part of the 3-qubit decomposition. I suggest this issue to be closed;
* the running time can be reduced sustantially by optimizing `cleanup_operations`. I am happy to submit my fix here. I suggest spending time to also fix `merge_single_qubit_gates_to_phased_x_and_z`, but this can be a different issue.

@NoureldinYosri @ACE07-Sev can you please have a look and let me know your thoughts?
@ACE07-Sev can you rerun your analysis and confirm the depth is better at the current package version?

**Comment by ACE07-Sev** [comment]
I implemented A.2 a few months ago for `quick` and it definitely improves depth. 

I'll have a look soon but in the meantime kindly check yours with mine.

**Comment by codrut3** [comment]
@ACE07-Sev as I explain in the comment A.2 is already implemented here, but it's not obvious, because it's part of the `n=3` base case. I ran my local analysis and didn't see any improvement from removing the `n=3` base case and using `n=2` and A.2. However, running the circuit through the optimizer substantially reduces depth: the issue is solved because the optimizer got better and is similar to the Qiskit one now.

If you have a colab or script to share, I'm happy to run it.

**Comment by ACE07-Sev** [comment]
@codrut3 I'm comparing with mine now. Can you run yours with transpile afterwards? I want to compare the depths with mine.

**Comment by codrut3** [comment]
Are these Qiskit numbers? The Cirq numbers are:
```
qubits = [3, 4, 5, 6, 7, 8]
cirq_package = [41, 203, 899, 3779, 15491, 62723]
```
Would you consider this to be close enough?

**Comment by ACE07-Sev** [comment]
Yes, those are the qiskit numbers (after transpile). Did you run transpile on yours as well?

**Comment by codrut3** [comment]
Yes, this is after the transpiler (I call it optimizer :)). Basically, I think the transpiler does the heavy work and so further improvements would be realized by improving the transpiler.

**Comment by ACE07-Sev** [comment]
I see. I mean, for small number of qubits sure, but when you go above 10 it's gonna make a big difference. Especially with practical use-cases given the deeper the circuit is the noisier it's going to get.

**Comment by ACE07-Sev** [comment]
I'll put this on my TODO list to properly go through the codebase ASAP. Right now my laptop is kaput and freezes every few minutes so can't properly go through it.

**Comment by ACE07-Sev** [comment]
I'll run it without transpile right now. You do the same please. We can then compare. If they match then LGTM, but otherwise we could really use that small improvement.

**Comment by codrut3** [comment]
> I'll run it without transpile right now. You do the same please. We can then compare. If they match then LGTM, but otherwise we could really use that small improvement.

My numbers are:
```
qubits = [3, 4, 5, 6, 7, 8]
cirq_package = [45, 219, 963, 4035, 16515, 66819]
with_my_change = [47, 221, 965, 4037, 16517, 66821]
```
> P.S : I get you on the laptop being slow hehe. God knows I have had code that ran for a week on my laptop. I would definitely try either using multi-threading to make the loops for each qubit to run in parallel (you'd have to use 3.14 with nogil to do this properly) or to get a CPU with really fast single-core speed. That works wonders.

Thank you! Using multi-threading: doesn't this require writing the shannon decomposition to be multi-threaded?

> I'm not sure if you're a community member like moi or you work at cirq, but if the latter, definitely try to run the code on sth like i7 14700k or above (the office should have a CPU like that). It can cut the runtime in 4 usually and make it easier to work on.

I work in a different part of Google. I contribute to cirq in my free time on my personal laptop. For now I'll try to stick to this :)

> By the way, before you start optimizing, please run a profiler to see what is taking the most amount of time. cProfiler is great for that (let me know if you'd like some assistance with this, I love profiling). You can even make a call graph png with gprof2dot to visualize it and make it easier to read (it also makes it easier to catch redundancies based on number of calls to a function/method).
> 
> Also, keep in mind python is not the best language for runtime speed. Hence why some packages have been porting the heavy-duty stuff to rust and only use python as the interface.

I used a profiler and it did help to narrow down but I'm still not sure of the root cause. I wrote an optimization for `cleanup_operations`, but it would be better to optimize `merge_single_qubit_gates_to_phased_x_and_z`. I'll try to use gprof2dot, thanks for the tip!
I know python goes only so far, but there is definitely some runtime that can be squeezed here :)

**Comment by ACE07-Sev** [comment]
Should be fine then. I checked the older messages and I had said 70k there without transpile so I guess it _could_ be transpile bringing it to 60k.

I'll have to properly check this to make sure. I think one way for you to be sure (I tried, I get frozen so can't run it myself, looking to get a new hardware soon to not be so useless like now) is to convert the cirq circuit you're getting before optimization to qiskit via qasm or via my package if you prefer that and then call transpile on both and then compare. That way if they don't match then you know for a fact that the decomposition are not the same. That should remove the effect of transpile from the comparison.

**Comment by ACE07-Sev** [comment]
No no what I meant by multiprocessing is to just use it for when you were generating the circuit depths for a list of qubit sizes. Basically since each iteration is independent then you can run them in parallel to save time.

Like instead of generating circuit for 1 qubit then 2 qubit then 3 qubit and so on you would instead generate all of them in parallel since they are independent tasks. That way it would take as long as the longest iteration. It would save considerable time.

I wouldn't recommend making cirq itself use multiprocessing given it's a very user-specific thing (some have very few cores or very slow clock speed so for them it would be slower than no multiprocessing due to the overhead of the communication between the threads).

**Comment by NoureldinYosri** [comment]
thanks @codrut3 and @ACE07-Sev for the analysis and discussion ... looking forward to your final verdict

@ACE07-Sev your optimized `cleanup_operations` will definitely be a great addition to Cirq, also what is the bottleneck in `merge_single_qubit_gates_to_phased_x_and_z`? 

@ACE07-Sev and @codrut3 I don't know if you are aware but we have a biweekly cirq-sync meeting, if you want to attend you can join the cirq-dev group https://groups.google.com/g/cirq-dev which will send you an invite to the meeting.

**Comment by codrut3** [comment]
Thanks @NoureldinYosri ! I'll come to the next Cirq Sync to discuss this. @ACE07-Sev would you be able to attend too? It's on Wednesday 28th May.

**Comment by ACE07-Sev** [comment]
So so sorry for the delay. I had some court issues and forgot about this. Sure, I'll try to be there.

**Comment by ACE07-Sev** [comment]
@codrut3 Did you get a chance to run the experiment I asked? That's the main thing I would like to see to finally provide a meaningful comparison. Would be great to have that before the biweekly meeting on 28th.

**Comment by ACE07-Sev** [comment]
@NoureldinYosri You are too kind. I have been a bit useless with my laptop being glitchy. I have put this issue on my TODO if it's still open by the time I get a new hardware.

**Comment by codrut3** [comment]
> I'll have to properly check this to make sure. I think one way for you to be sure (I tried, I get frozen so can't run it myself, looking to get a new hardware soon to not be so useless like now) is to convert the cirq circuit you're getting before optimization to qiskit via qasm or via my package if you prefer that and then call transpile on both and then compare. That way if they don't match then you know for a fact that the decomposition are not the same. That should remove the effect of transpile from the comparison.

Thank you for the suggestion! I did this. More precisely, I ran the Cirq Shannon decomposition, converted to qasm and then ran Qiskit transpiler with basis gates u3 and cz. Then I ran Cirq transpiler with the same target gateset.

Both transpilers give the same depth!

For the current Cirq package:
```
qubits = [3, 4, 5, 6, 7, 8]
cirq_depth_after_transpiler = [41, 203, 899, 3779, 15491, 62723]
qiskit_depth_after_transpiler = [41, 203, 899, 3779, 15491, 62723]
```

With my change that implements A.2:
```
qubits = [3, 4, 5, 6, 7, 8]
cirq_depth_after_transpiler = [43, 205, 901, 3781, 15493, 62725]
qiskit_depth_after_transpiler = [43, 205, 901, 3781, 15492, 62724]
```

Here there is a `-1` difference. 

One thing to note is that the Cirq transpiler creates more u3 gates (but this doesn't change the depth). The cz gate numbers after the transpiler match.

Do I interpret this correctly that it's not the transpiler, but the Cirq Shannon decomposition that still has some inefficiency that causes higher depth? But it can't be A.2, because I implemented it and it doesn't change things.
